src/crimson/osd/replicated_recovery_backend.h

```diff
@@ -47,16 +47,37 @@ class ReplicatedRecoveryBackend : public RecoveryBackend {
     Ref<MOSDPGRecoveryDelete> m);
   interruptible_future<> handle_recovery_delete_reply(
     Ref<MOSDPGRecoveryDeleteReply> m);
+  interruptible_future<PushOp> prep_push_to_replica(
+    const hobject_t& soid,
+    eversion_t need,
+    std::pair<pg_shard_t, pg_missing_t> pg_shard);
+  void calc_clone_subsets(
+    SnapSet& snapset, const hobject_t& soid,
+    const pg_missing_t& missing,
+    const hobject_t &last_backfill,
+    interval_set<uint64_t>& data_subset,
+    std::map<hobject_t, interval_set<uint64_t>>& clone_subsets);
+  void calc_head_subsets(
+    crimson::osd::ObjectContextRef obc,
+    SnapSet& snapset,
+    const hobject_t& head,
+    const pg_missing_t& missing,
+    const hobject_t &last_backfill,
+    interval_set<uint64_t>& data_subset,
+    std::map<hobject_t, interval_set<uint64_t>>& clone_subsets);
   interruptible_future<PushOp> prep_push(
     const hobject_t& soid,
     eversion_t need,
-    pg_shard_t pg_shard);
+    pg_shard_t pg_shard,
+    interval_set<uint64_t> &data_subset,
+    std::map<hobject_t, interval_set<uint64_t>>& clone_subsets);
   void prepare_pull(
-    PullOp& po,
-    PullInfo& pi,
+    crimson::osd::ObjectContextRef obc,
+    PullOp& pull_op,
+    pull_info_t& pull_info,
     const hobject_t& soid,
     eversion_t need);
-  std::vector<pg_shard_t> get_shards_to_push(
+  std::map<pg_shard_t, pg_missing_t> get_shards_to_push(
```

Comment:
> Alternately, it could be std::map<pg_shard_t, pg_missing_t> including just the entry for the object in question?

`get_shards_to_push()` is returning a vector (`shards`) of `pg_shard_t` according to the following logic:
```
auto shard_missing = pg.get_shard_missing().find(peer);
if (shard_missing->second.is_missing(soid))
  shards.push_back(shard_missing->first);
```
Each of this vector's iterators is passed to `prep_push` immediately (unconditionally) afterwards:
```
const auto& missing = pg.get_shard_missing().find(pg_shard)->second;
const auto it = missing.get_items().find(soid);
assert(it != missing.get_items().end());
```

Instead of pushing only `->first`, and then finding the `pg_shard` again to get `->second`,
`get_shards_to_push()` is updated to return a map and `interruptor::parallel_for_each` will pass pairs to `prep_push` (commit: `Introduce prep_push_to_replica()`).

In short,  we avoid finding the missing peer and asserting it's existence (again) later on in prep_push_to_replica.

Reply:
Right, the problem is that it means copying the pg_missing_t structures, which can be big and will involve a lot of allocations.  That's a lot of unnecessary overhead, even for peering.

Reply:
Ack, will be fixed.